vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java

```diff
@@ -159,49 +161,67 @@ public Set<T> getSelectedItems() {
                 .unmodifiableSet(new LinkedHashSet<>(selected.values()));
     }
 
+    /**
+     * Returns an unmodifiable view of the selected item ids.
+     * <p>
+     * The returned Set may be a direct view of the internal data structures of
+     * this class. A defensive copy should be made by callers when iterating
+     * over this Set and modifying the selection during iteration to avoid
+     * ConcurrentModificationExceptions.
+     *
+     * @return An unmodifiable view of the selected item ids. Updates in the
+     *         selection may or may not be directly reflected in the Set.
+     */
+    protected Set<Object> getSelectedItemIds() {
+        return Collections.unmodifiableSet(this.selected.keySet());
+    }
+
     @Override
     public Optional<T> getFirstSelectedItem() {
         return selected.values().stream().findFirst();
     }
 
     @Override
     public void select(T item) {
-        if (isSelected(item)) {
+        Object itemId = getItemId(item);
+        if (isSelectedItemId(item)) {
             return;
         }
-        Set<T> selected = new HashSet<>();
+        Map<Object, T> selected = new HashMap<>();
```

Comment:
Should it be LinkedHashMap everywhere to make sure the selection order is preserved?

Reply:
Refactored.